chain-core/src/state/tendermint.rs

```diff
@@ -13,20 +13,53 @@ use std::prelude::v1::{String, ToString, Vec};
 /// TODO: u64?
 pub type BlockHeight = i64;
 
+/// ed25519 public key size
+pub const PUBLIC_KEY_SIZE: usize = 32;
+
 /// The protobuf structure currently has "String" to denote the type / length
 /// and variable length byte array. In this internal representation,
 /// it's desirable to keep it restricted and compact. (TM should be encoding using the compressed form.)
 #[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Clone, Encode, Decode)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(tag = "type", content = "value"))]
 pub enum TendermintValidatorPubKey {
-    Ed25519([u8; 32]),
+    #[cfg_attr(
+        feature = "serde",
+        serde(
+            rename = "tendermint/PubKeyEd25519",
+            serialize_with = "serialize_ed25519_base64",
+            deserialize_with = "deserialize_ed25519_base64"
+        )
+    )]
+    Ed25519([u8; PUBLIC_KEY_SIZE]),
     // there's PubKeySecp256k1, but https://tendermint.com/docs/spec/abci/apps.html#validator-updates
     // "The pub_key currently supports only one type:"
     // "type = "ed25519" anddata = <raw 32-byte public key>`"
     // there's also PubKeyMultisigThreshold, but that probably wouldn't be used for individual nodes / validators
     // TODO: some other schemes when they are added in TM?
 }
 
+/// Serialize the bytes of an Ed25519 public key as Base64. Used for serializing JSON
+#[cfg(feature = "base64")]
+fn serialize_ed25519_base64<S>(pk: &[u8], serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    base64::encode(pk).serialize(serializer)
+}
+
+#[cfg(feature = "base64")]
+fn deserialize_ed25519_base64<'de, D>(deserializer: D) -> Result<[u8; PUBLIC_KEY_SIZE], D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let bytes = base64::decode(String::deserialize(deserializer)?.as_bytes())
+        .map_err(|e| D::Error::custom(format!("{}", e)))?;
+    let mut result = [0u8; PUBLIC_KEY_SIZE];
+    result.copy_from_slice(&bytes);
```

Comment:
`copy_from_slice` will panic if the lengths don't match, so ideally before this, there should be a check on `bytes.len() == PUBLIC_KEY_SIZE`

Reply:
also perhaps just deserialise the string and then call `TendermintValidatorPubKey::from_base64` ?

Reply:
Fixed, also brings in `thiserror` crate, looks like a decent way to handle error type hierarchies.